### PR TITLE
Avoid in-place operations when disabling the smart_inference_mode to compute the gradients.

### PR DESCRIPTION
**Pull request by Bimo99B9**
_I have read the CLA Document and I hereby sign the CLA_

---

This PR is related to the issue #6211.

The gradient computation was giving errors when disablig the `smart_inference_mode` because of in-place operations:

image

Changes made:

1. Scale the boxes without modifying the original pred tensor directly, and create a new tensor by combining scaled boxes with the rest of pred to avoid in-place operations.
2. In `ops.py`, the clipped values are saved in `clipped_boxes` to avoid modifying the original boxes. This method, `clip_boxes`, already had a warning in a comment because inplace .clamp, which causes bugs.

I was able to calculate the gradients with these changes, feel free to ask for further explanations.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
No code changes in this pull request.

### 📊 Key Changes
- There are no changes to the code.

### 🎯 Purpose & Impact
- This PR does not introduce any new features, bug fixes, or updates. Therefore, there is no impact on users or the functionality of the project at this time.

### CODE REVIEW

**Review by Burhan-Q — COMMENTED** [review]
What are the measured differences with respect to memory usage when using these changes?

**Comment by Bimo99B9** [comment]
Thanks for the review @Burhan-Q

I could not measure the differences in memory usage because when I disabled smart_inference_mode and computed the loss function (objectness_score), the model crashed.

If you are referring to the memory usage of the model with smart_inference_mode enabled, after applying these changes vs. before applying them, I could measure the memory usage of the two versions.

However, if you could give a hint on how to specifically do the memory profiling to measure the influence of these changes, I will be happy to do it, but I have never done these tests before, so I am not very familiar with them.

**Comment by Burhan-Q** [comment]
@Bimo99B9 I was referring to testing with your proposed changes keeping `smart_inference_mode` enabled. I can point you to a more robust profiling method if you like, but I was merely asking for monitoring using a simple hardware/resource monitoring solution. Anything that would provide you a plot to show any differences in an A/B test from a few runs would be easiest.

**Comment by Bimo99B9** [comment]
@Burhan-Q I ran two simple tests using the model to predict on a sample image and there was no difference in memory usage in the results.

I used `memory_profiler` to measure the memory used for prediction. 

image
image

I ran the same notebook with the original ultralytics repository installed, and with the one with these changes applied

To confirm that I was testing the modified model, I added some prints.
In the modified test, the maximum memory usage was 1338.3203125 MiB, and in the original one, the maximum model memory usage was 1340.9140625 MiB. I did a few runs manually and the values fluctuated between 1337 and 1341 MiB for both versions.

I understand that this is a minimal change and shouldn't have much of an impact on memory usage, but please let me know if you want me to do more runs or test something else where these changes might have an impact.

**Comment by Burhan-Q** [comment]
@Bimo99B9 thanks for sharing your results! These are a good to have and it's nice to see the system memory isn't impacted too much. The other thing that would be helpful to see is the difference (if any) for GPU memory usage, since when CUDA is available it would be loaded into GPU memory. It really helps to understand something like this before we start running other checks that are needed prior to a possible merge.

**Comment by Bimo99B9** [comment]
Thanks for your comments @Burhan-Q.

I ran some tests for GPU memory usage when using the model to predict an image.

This was my function to test the changes, please make sure that it tests what you are trying to validate the GPU memory usage on:

```python
import torch
import matplotlib.pyplot as plt
from ultralytics import YOLO

if not torch.cuda.is_available():
    raise SystemError("CUDA is not available. GPU memory tracking requires CUDA.")

def test_model():
    model = YOLO("yolov8l.pt").cuda()
    torch.cuda.synchronize()
    
    start_mem_allocated = torch.cuda.memory_allocated()
    start_mem_reserved = torch.cuda.memory_reserved()
    
    results = model("https://ultralytics.com/images/bus.jpg")
    torch.cuda.synchronize()
    
    end_mem_allocated = torch.cuda.memory_allocated()
    end_mem_reserved = torch.cuda.memory_reserved()
    
    delta_mem_allocated = (end_mem_allocated - start_mem_allocated) / (1024 ** 2)
    delta_mem_reserved = (end_mem_reserved - start_mem_reserved) / (1024 ** 2)
    
    del model, results
    torch.cuda.empty_cache()
    
    return delta_mem_allocated, delta_mem_reserved

num_iterations = 20
allocations = []
reservations = []

for i in range(num_iterations):
    delta_allocated, delta_reserved = test_model()
    allocations.append(delta_allocated)
    reservations.append(delta_reserved)
```

These are the results, plotted side by side. The modified one is on the left, and the original one is on the right:

image

**Review by Burhan-Q — APPROVED** [review]
Thanks for the testing! I think that this seems reasonable and is well prepared for the final review! I don't know when or if it will be reviewed for certain, but I think your updates and information will certainly increase the odds! 🚀

**Comment by glenn-jocher** [comment]
@Bimo99B9 absolutely, happy to help further if needed! 🌟 If you have any more questions or need assistance with anything else, feel free to reach out. Keep up the great work!

**Comment by Bimo99B9** [comment]
Hi, any update on this?

**Comment by Burhan-Q** [comment]
Hey @Bimo99B9 I know it's been awhile, but it looks like some of the `pytests` are now failing and there might be a merge conflict that needs to be fixed as well. If you can fix these, you can ping me and I'll shoot over a reminder for a final review but we have so much work in so many places, that I can't say for certain when it might happen. I think even I have 4 of 5 open PRs that are waiting on a final review 😅

**Comment by Bimo99B9** [comment]
@Burhan-Q I just updated my fork, the changes work with the latest versions (I tested them these days). Let me know if you need anything else to merge this in

**Comment by glenn-jocher** [comment]
Thank you for this careful work tracking down the in-place operations that broke gradient computation with `smart_inference_mode` disabled (issue #6211), and for the thorough memory profiling you ran. Since this PR was opened, the relevant code has changed substantially: `ops.clip_boxes` now intentionally keeps in-place `clamp_` for performance with dedicated Apple MPS-bug handling (see PR #21878), and `detect/predict.py` has been refactored into new `construct_results`/`construct_result` methods, so the diff no longer applies and now conflicts with main. Because the current design deliberately favors the in-place approach, we're going to close this one, but we'd genuinely welcome a fresh look if you'd like to revisit gradient support against the current code. Thanks again for contributing.
